src/librustdoc/html/highlight.rs

```diff
@@ -119,336 +83,262 @@ enum Class {
-    fn write_source<W: Writer>(&mut self, out: &mut W) -> Result<(), HighlightError> {
-        loop {
-            let mut next = self.try_next_token()?;
-            if next == token::Eof {
-                break;
-            }
-
-            // Glue any tokens that need to be glued.
-            if let Some(joint) = next.glue(self.peek()?) {
-                next = joint;
-                let _ = self.try_next_token()?;
+    /// possibly giving it an HTML span with a class specifying what flavor of
+    /// token is used.
+    fn highlight(mut self, sink: &mut dyn FnMut(Highlight<'a>)) {
+        with_default_session_globals(|| {
+            while let Some((token, text)) = self.tokens.next() {
+                self.advance(token, text, sink);
             }
-
-            self.write_token(out, next)?;
-        }
-
-        Ok(())
+        })
     }
 
-    // Handles an individual token from the lexer.
-    fn write_token<W: Writer>(&mut self, out: &mut W, token: Token) -> Result<(), HighlightError> {
-        let klass = match token.kind {
-            token::Shebang(s) => {
-                out.string(Escape(&s.as_str()), Class::None)?;
-                return Ok(());
-            }
-
-            token::Whitespace | token::Unknown(..) => Class::None,
-            token::Comment => Class::Comment,
-            token::DocComment(..) => Class::DocComment,
-
-            // If this '&' or '*' token is followed by a non-whitespace token, assume that it's the
-            // reference or dereference operator or a reference or pointer type, instead of the
-            // bit-and or multiplication operator.
-            token::BinOp(token::And | token::Star) if self.peek()? != &token::Whitespace => {
-                Class::RefKeyWord
+    /// Single step of highlighting. This will classify `token`, but maybe also
+    /// a couple of following ones as well.
+    fn advance(&mut self, token: TokenKind, text: &'a str, sink: &mut dyn FnMut(Highlight<'a>)) {
+        let lookahead = self.peek();
+        let class = match token {
+            TokenKind::Whitespace => Class::None,
+            TokenKind::LineComment { doc_style } | TokenKind::BlockComment { doc_style, .. } => {
+                if doc_style.is_some() {
+                    Class::DocComment
+                } else {
+                    Class::Comment
+                }
             }
-
             // Consider this as part of a macro invocation if there was a
             // leading identifier.
-            token::Not if self.in_macro => {
+            TokenKind::Bang if self.in_macro => {
                 self.in_macro = false;
                 Class::Macro
             }
 
+            // Assume that '&' or '*' is the reference or dereference operator
+            // or a reference or pointer type. Unless, of course, it looks like
+            // an operator: `&&` or `* `.
+            TokenKind::Star => match lookahead {
+                Some(TokenKind::Whitespace) => Class::Op,
```

Comment:
Maybe mention that this is multiplication, I was confused and thought you meant the dereference operator at first.

Also, this seems really fragile and will break on things like `1*2`. Is there a reason we're not using a full parser? It would be lot more work so definitely a follow up PR, but otherwise I don't see a way to avoid bugs here.

Reply:
The original tradeoff was likely simplicity and performance over correctness. IMO, I don't think the rustdoc highlighting needs to be perfect, as long as it's reasonably correct for well-formatted, idiomatic code.

There's obviously edge cases like `&&foo` that are incorrectly highlighted by the lexer, but mis-highlighting `1*2` doesn't seem like a problem to me when `1 * 2` is highlighted correctly.